src/libcore/array/mod.rs

```diff
@@ -387,6 +387,84 @@ where
     }
 }
 
+/// Non-zero numbers
+///
+/// We require following from this trait:
+///  - [(); 0] does **not** implement NonZero.
+///  - [(); N] where N != 0 **implements** NonZero.
+///  As side effect of selected approach, this trait is also implemented
+///  for all other types, but it does not matter.
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+pub auto trait NonZero {}
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+impl !NonZero for [(); 0] {}
+
+// This is implementation for all non-zero-sized arrays
+#[stable(since = "1.4.0", feature = "array_default")]
+impl<T: Default, const N: usize> Default for [T; N]
+where
+    [(); N]: NonZero,
+{
+    #[inline]
+    fn default() -> [T; N] {
+        use crate::mem::MaybeUninit;
+        // invariant: first `init` items are initialized
+        struct Wrapper<T, const N: usize> {
+            data: MaybeUninit<[T; N]>,
+            init: usize,
+        }
+
+        impl<T, const N: usize> Drop for Wrapper<T, N> {
+            #[inline]
+            fn drop(&mut self) {
+                debug_assert!(self.init <= N);
+
+                let ptr = self.data.as_mut_ptr() as *mut T;
+                let initialized_part = crate::ptr::slice_from_raw_parts_mut(ptr, self.init);
+                // SAFETY: this raw slice will contain only initialized objects
+                // that's why, it is allowed to drop it.
+                unsafe {
+                    crate::ptr::drop_in_place(initialized_part);
+                }
+            }
+        }
+
+        let mut w = Wrapper::<T, N> { data: MaybeUninit::uninit(), init: 0 };
+        let array_pointer = w.data.as_mut_ptr() as *mut T;
+        for i in 0..N {
+            // This does not work where N is bigger than isize::max.
+            // FIXME: do we care?
+            assert!(N <= isize::MAX as usize);
+            // SAFETY: i < N <= isize::MAX, so add() call is valid.
+            unsafe {
+                let elem_ptr = array_pointer.add(i);
```

Comment:
You can keep pointer instead of an integer and `.add(1)` it, that might optimize better.

In fact, you should be able to get a `&mut [MaybeUninit<T>; N]` and "just" iterate that, but I'm not sure that's faster (or safer)?

cc @RalfJung

Reply:
Yeah sometimes `[MaybeUninit<T>; N]` is easier to work with than the `MaybeUninit<[T; N]>` you are using (and the two are freely transmutable).